### ISSUE TITLE
Change to rollingUpdate values in kustomize 3.8.0

### ISSUE THREAD

**Issue by robinjmurphy** [kubernetes-sigs/kustomize#2790]
Hello 👋 We've recently upgraded from Kustomise `3.7.0` to `3.8.0` and have seen two changes in behaviour:

1. `io.k8s.api.apps.v1.RollingUpdateDeployment.maxSurge` and `io.k8s.api.apps.v1.RollingUpdateDeployment.maxUnavailable` values in patches that are defined as integers are now converted to strings (resulting in an invalid template)
2. The `$retainKeys` directive in patches appears to be included in generated templates, whereas it wasn't before

I don't have much context on the kustomize tool, but I'm hoping to figure out whether this is a bug in the tool that was introduced in `3.8.0` or whether this is correct and I should update my patch files.

Steps to reproduce...

```
mkdir -p kustomize-test
touch kustomize-test/{kustomization,deployment,migration}.yaml
```

```yaml
# kustomization.yaml
namespace: foo
patchesStrategicMerge:
- migration.yaml
resources:
- deployment.yaml
```

```yaml
# deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: foo
spec:
  selector:
    matchLabels:
      app: nginx
  strategy:
    rollingUpdate:
      maxSurge: 30%
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: foo-ctr
        image: nginx
```

```yaml
# migration.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: foo
spec:
  strategy:
    $retainKeys:
    - rollingUpdate
    rollingUpdate:
      maxSurge: 1
```

Output with `3.7.0`:

```
❯ kustomize-370 build kustomize-test
apiVersion: apps/v1
kind: Deployment
metadata:
  name: foo
  namespace: foo
spec:
  selector:
    matchLabels:
      app: nginx
  strategy:
    rollingUpdate:
      maxSurge: 1
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        name: foo-ctr
```

Output with `3.8.0`:

```
❯ kustomize-380 build kustomize-test
apiVersion: apps/v1
kind: Deployment
metadata:
  name: foo
  namespace: foo
spec:
  selector:
    matchLabels:
      app: nginx
  strategy:
    $retainKeys:
    - rollingUpdate
    rollingUpdate:
      maxSurge: "1"
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        name: foo-ctr
```

Diff:

```diff
--- 3.7.0	2020-07-30 18:21:45.000000000 +0100
+++ 3.8.0	2020-07-30 18:21:48.000000000 +0100
@@ -8,8 +8,10 @@
     matchLabels:
       app: nginx
   strategy:
+    $retainKeys:
+    - rollingUpdate
     rollingUpdate:
-      maxSurge: 1
+      maxSurge: "1"
   template:
     metadata:
       labels:
```

**Comment by robinjmurphy** [comment]
Have tested in `3.8.1` and the results are the same as `3.8.0`.

**Comment by robinjmurphy** [comment]
The reason this results in an invalid template is because a string value is assumed to be a percentage and these values do not include the `%` suffix.

**Comment by acesaro** [comment]
Any workaround for this?

**Comment by acesaro** [comment]
> Any workaround for this?

The workaround we found was to ensure that the values in base are of the same data type as your patches.  For example, we had "string" placeholder values in base while we patched them with "integer" actual desired values in our patches.  Prior to 3.8.0, this actually kept the data type of the patch, regardless of what was in base.  After thinking about it more, this new behavior is more desired from our perspective, as it enforces data type consistency.
